File: iMMPI/iMMPI/JSONTestRecordProxy.m
//
//  JSONTestRecordProxy.m
//  iMMPI
//
//  Created by Egor Chiglintsev on 13/02/14.
//  Copyright (c) 2014 Egor Chiglintsev. All rights reserved.
//

#if (!__has_feature(objc_arc))
#error "This file should be compiled with ARC support"
#endif

#import "JSONTestRecordProxy.h"
#import "JSONTestRecordSerialization.h"


#pragma mark -
#pragma mark JSONTestRecordProxy private

@interface JSONTestRecordProxy()
{
@private
    id<TestRecordProtocol> _record;
}

@end


#pragma mark -
#pragma mark JSONTestRecordProxy implementation

@implementation JSONTestRecordProxy

#pragma mark -
#pragma mark properties

- (id<PersonProtocol>) person
{
    [self loadRecordIfNeeded];
    return _record.person;
}


- (void) setPerson:(id<PersonProtocol>)person
{
    [self loadRecordIfNeeded];
    _record.person = person;
}


- (id<TestAnswersProtocol>) testAnswers
{
    [self loadRecordIfNeeded];
    return _record.testAnswers;
}


- (void) setTestAnswers: (id<TestAnswersProtocol>) testAnswers
{
    [self loadRecordIfNeeded];
    _record.testAnswers = testAnswers;
}


- (NSString *) personName
{
    if (_record != nil)
    {
        return _record.person.name;
    }
    else
    {
        return _personName;
    }
}


- (NSDate *) date
{
    if (_record != nil)
    {
        return _record.date;
    }
    else if (_date != nil)
    {
        return _date;
    }
    else
    {
        [self loadRecordIfNeeded];
        return _record.date;
    }
}


#pragma mark -
#pragma mark methods

+ (id) proxyForRecord: (id<TestRecordProtocol>) record
         withFileName: (NSString *) fileName
          inDirectory: (NSString *) directory
{
    return [[[self class] alloc] initWithRecord: record
                                   withFileName: fileName
                                    inDirectory: directory];
}


#pragma mark -
#pragma mark initialization

- (id) initWithRecord: (id<TestRecordProtocol>) record
         withFileName: (NSString *) fileName
          inDirectory: (NSString *) directory
{
    self = [super init];
    
    if (self != nil)
    {
        _fileName   = [fileName  copy];
        _directory  = [directory copy];
        _personName = record.person.name;
        _date       = record.date;
        _record     = record;
    }
    return self;
}


#pragma mark -
#pragma mark private

- (void) loadRecordIfNeeded
{
    if (_record == nil)
    {
        NSLog(@"Proxy loading record for %@", _personName);
        
        NSArray  *directories = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
        NSString *storedRecordsPath = [[directories lastObject] stringByAppendingPathComponent:
                                       _directory];
        
        NSString *recordPath = [storedRecordsPath stringByAppendingPathComponent: _fileName];
        NSData   *recordData = [NSData dataWithContentsOfFile: recordPath];
        
        _record = [JSONTestRecordSerialization testRecordFromData: recordData];
    }
}


#pragma mark -
#pragma mark debugging

- (NSString *) debugDescription
{
    return [NSString stringWithFormat:
            @"%@: %@ (%@)", [super debugDescription], _personName, _date];
}

@end

File: iMMPI/iMMPI/JSONTestRecordProxy.swift
import Foundation

final class JSONTestRecordProxy: NSObject {
    let fileName: String
    let directory: String

    let serialization = JSONTestRecordSerialization()

    init(fileName: String, directory: String) {
        self.fileName = fileName
        self.directory = directory
    }

    fileprivate var _record: TestRecordProtocol?
    fileprivate var _personName = ""
    fileprivate var _date: Date?
}


extension JSONTestRecordProxy {
    override var description: String {
        return "\(super.description): \(personName), (\(date))"
    }


    convenience init(record: TestRecordProtocol, fileName: String, directory: String) {
        self.init(fileName: fileName, directory: directory)
        self._record = record
    }
}


extension JSONTestRecordProxy: TestRecordProtocol {
    var person: PersonProtocol {
        get {
            return record.person
        }

        set {
            record.person = newValue
        }
    }


    var testAnswers: TestAnswers {
        get {
            return record.testAnswers
        }

        set {
            record.testAnswers = newValue
        }
    }


    var personName: String {
        get {
            if let record = _record {
                return record.personName
            }
            else {
                return _personName
            }
        }

        set {
            _personName = newValue
        }
    }


    var date: Date {
        get {
            if let record = _record {
                return record.date
            }
            else if let date = _date {
                return date
            }
            else {
                return record.date
            }
        }


        set {
            _date = newValue
        }
    }
}


extension JSONTestRecordProxy {
    fileprivate var record: TestRecordProtocol {
        if _record == nil {
            loadRecord()
        }

        return _record!
    }


    fileprivate func loadRecord() {
        var loadedRecord: TestRecordProtocol = TestRecord()

        defer {
            _record = loadedRecord
        }

        NSLog("Proxy loading record for \(personName)")

        guard let documentsUrl = FileManager().urls(for: .documentDirectory, in: .userDomainMask).first else {
            assertionFailure("Failed locating documents directory")
            return
        }


        let recordsDirectory = documentsUrl.appendingPathComponent(directory)
        let recordPath = recordsDirectory.appendingPathComponent(fileName)

        let data: Data

        do {
            try data = Data(contentsOf: recordPath)
        }
        catch let error {
            assertionFailure("Error reading data: \(error)")
            return
        }

        loadedRecord = serialization.decode(data) ?? loadedRecord
    }
}
